src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -262,11 +273,129 @@ std::vector<datalog::Rule> LoweringToDatalogPass::QueriesToDLIR(
+                                       "Principal"))}));
+  // Attributes have a principal that the attribute is applied to
+  // as an additional parameter at the beginning of the param list.
+  // Transform the attributes
+  std::vector<datalog::RelationDeclaration> transformed_declarations =
+      utils::MapIter<datalog::RelationDeclaration>(
+          base_declarations, [](datalog::RelationDeclaration declaration) {
+            if (declaration.is_attribute()) {
+              std::vector<datalog::Argument> transformed_arguments = {
+                  datalog::Argument("attribute_prin",
+                                    datalog::ArgumentType(
+                                        datalog::ArgumentType::Kind::kPrincipal,
+                                        "Principal"))};
+              transformed_arguments.insert(transformed_arguments.end(),
+                                           declaration.arguments().begin(),
+                                           declaration.arguments().end());
+              return datalog::RelationDeclaration(declaration.relation_name(),
+                                                  false, transformed_arguments);
+            } else {
+              return declaration;
+            }
+          });
+  // Produce a mapping from predicate names to predicate typings
+  // (where a predicate typing is the same as a predicate relation declaration)
+  absl::flat_hash_map<std::string_view, datalog::RelationDeclaration>
+      type_environment;
+  for (datalog::RelationDeclaration declaration : transformed_declarations) {
+    type_environment.insert({declaration.relation_name(), declaration});
+  }
+
+  // Generate relation declarations for delegated predicates with
+  // each encountered delegation depth
+  std::vector<datalog::RelationDeclaration> can_say_declarations;
+  for (const auto& can_say_depth_pair : can_say_depth) {
+    auto type_iterator = type_environment.find(can_say_depth_pair.first);
+
+    if ((can_say_depth_pair.second > 0) &&
+        (type_iterator != type_environment.end())) {
+      datalog::RelationDeclaration declaration = type_iterator->second;
+      for (uint64_t i = 1; i < can_say_depth_pair.second + 1; ++i) {
+        std::vector<datalog::Argument> argument = {datalog::Argument(
+            absl::StrCat("delegatee", i),
+            datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                  "Principal"))};
+        argument.insert(argument.end(), declaration.arguments().begin(),
+                        declaration.arguments().end());
+        can_say_declarations.push_back(datalog::RelationDeclaration(
+            absl::StrCat("canSay_", declaration.relation_name()), false,
+            argument));
+        declaration = datalog::RelationDeclaration(
+            absl::StrCat("canSay_", declaration.relation_name()), false,
+            argument);
+      }
```

Comment:
This loop is a bit hard to follow. Can you give me an example of what is the result of the transition here is?

Reply:
Yes. That will makes things simpler. I will do that!